src/mpid/ch4/shm/shmam/eager/fbox/fbox_init.h

```diff
@@ -0,0 +1,168 @@
+                        MPIDU_shm_seg_info_ptr_t,
+                        MPIDI_SHMAM_eager_fbox_control_global.num_seg *
+                        sizeof(MPIDU_shm_seg_info_t), mpi_errno, "mem_region segments");
+
+    /* Request fastboxes region */
+
+    mpi_errno =
+        MPIDU_shm_seg_alloc(num_local * num_local * sizeof(MPIDI_SHMAM_fastbox_t),
+                            (void **) &fastboxes_p);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Request shared collective barrier vars region */
+
+    mpi_errno = MPIDU_shm_seg_alloc(MAX(sizeof(MPIDU_shm_barrier_t), MPIDU_SHM_CACHE_LINE_LEN),
+                                    (void **) &MPIDI_SHMAM_eager_fbox_control_global.
+                                    barrier_region);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Actually allocate the segment and assign regions to the pointers */
+
+    mpi_errno = MPIDU_shm_seg_commit(&MPIDI_SHMAM_eager_fbox_control_global.memory,
+                                     &MPIDI_SHMAM_eager_fbox_control_global.barrier,
+                                     num_local, grank, 0, rank);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* Allocate table of pointers to fastboxes */
+
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+    MPIR_CHKPMEM_MALLOC(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.out,
+                        MPIDI_SHMAM_fastbox_t **, num_local * sizeof(MPIDI_SHMAM_fastbox_t *),
+                        mpi_errno, "fastboxes");
+
+#define MPIDI_SHMAM_MAILBOX_INDEX(sender, receiver) ((num_local) * (sender) + (receiver))
+
+    /* Fill in fbox tables */
+
+    for (i = 0; i < num_local; i++) {
+        MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in[i] =
+            &fastboxes_p[MPIDI_SHMAM_MAILBOX_INDEX(i, grank)];
+        MPIDI_SHMAM_eager_fbox_control_global.mailboxes.out[i] =
+            &fastboxes_p[MPIDI_SHMAM_MAILBOX_INDEX(grank, i)];
+
+        memset(MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in[i], 0,
+               sizeof(MPIDI_SHMAM_fastbox_t));
+
+        MPIDI_SHMAM_eager_fbox_control_global.mailboxes.in[i]->flag = 0;
+    }
+
+    /* Run local procs barrier */
+
+    mpi_errno = MPIDU_shm_barrier(MPIDI_SHMAM_eager_fbox_control_global.barrier, num_local);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+#undef MPIDI_SHMAM_MAILBOX_INDEX
+
+    MPIR_CHKPMEM_COMMIT();
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_SHM_INIT);
+    return mpi_errno;
+  fn_fail:
+    /* --BEGIN ERROR HANDLING-- */
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+    /* --END ERROR HANDLING-- */
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_SHMAM_eager_finalize
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHMAM_eager_finalize()
```

Comment:
The namespace is rather weird here.  The API exposed to the glue layer would be MPIDI_SHMAM_.  Is "eager_finalize" meant to be exposed to the glue layer or internal?  If it is internal, shouldn't the naming be MPIDI_SHMAMI_?